File: picasso-extensions/src/main/java/io/github/hendraanggrian/picasso/transformation/CropSquareTransformer.java
package io.github.hendraanggrian.picasso.transformation;

import android.graphics.Bitmap;
import android.support.annotation.NonNull;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
class CropSquareTransformer extends Transformer {

    @NonNull
    @Override
    public Bitmap transform(@NonNull Bitmap source, boolean recycleSource) {
        final int size = Math.min(source.getWidth(), source.getHeight());
        final Bitmap target = Bitmap.createBitmap(source,
                (source.getWidth() - size) / 2,
                (source.getHeight() - size) / 2,
                size,
                size);
        if (source != target && recycleSource)
            source.recycle();
        return target;
    }

    @Override
    public String key() {
        return Key.newInstance(this).toString();
    }
}

File: picasso-utils/src/com/hendraanggrian/picasso/transformation/CropSquareTransformer.kt
package com.hendraanggrian.picasso.transformation

import android.graphics.Bitmap
import com.squareup.picasso.Transformation

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
internal class CropSquareTransformer : Transformation {

    override fun transform(source: Bitmap): Bitmap {
        val size = Math.min(source.width, source.height)
        val width = (source.width - size) / 2
        val height = (source.height - size) / 2
        val target = Bitmap.createBitmap(source, width, height, size, size)
        if (target != source) {
            source.recycle()
        }
        return target
    }

    override fun key() = "CropSquareTransformer"
}